mitodl/mit-open
# Learning Resource Card

### What are the relevant tickets?

[Learning Resource Cards #4356](https://github.com/mitodl/hq/issues/4356)

### Description (What does it do?)

Implements the Learning Resource Card "card" variants, sizes "4bloc" and "6bloc" in Figma.

https://www.figma.com/design/Eux3guSenAFVvNHGi1Y9Wm/MIT-Design-System?node-id=827-6168&m=dev

The "list" variants are to follow in subsequent PR.

Provides a base `Card` component and `LearningResourceCard` in `ol-components`. The base `Card` is also now in use by the cards in the Homepage "News and Events" section.


### Screenshots (if appropriate):

<img width="335" alt="image" src="https://github.com/mitodl/mit-open/assets/939376/8abe82c2-6ee9-4555-86aa-4fa04cff4c91">

<img width="335" alt="image" src="https://github.com/mitodl/mit-open/assets/939376/f7ccc92a-455b-4a80-86bb-54fd3f820fc6">


### How can this be tested?

This has been applied to the homepage carousels in the "Featured Courses" (medium/4bloc variant) and "Media" (small/6bloc variant) sections.

- Load the homepage and confirm the card match the designs in Figma.
  - The title should link to the resource URL.
  - Resources with a next start date should display the start date in the footer.
    - The best run start date is used if not on the parent resource.
    - OCW courses show "As taught in: \<semester\> \<year\>".
  - Certificated resources should display a "Certificate" badge.
  - Authenticated learning path editors should have an action button that opens the "add to learning path" dialog.
  - Authenticated users should have an action button that opens the "add to user list" dialog.
  - As resources are loading, the skeleton load state should be displayed and broadly reflect the loaded contents.

- A LearningResourceCard component has also been added to `ol-components` and should render correctly in Storybook (`yarn run storybook` and navigate to http://localhost:6006/?path=/story/ol-components-learningresourcecard--course).

- A base Card component has also been added to `ol-components` and should render correctly in Storybook (small and medium variants) - http://localhost:6006/?path=/story/ol-components-card--medium.

- The base Card component

### Additional Context

Developer notes:

- This adds a base Card component to ol-components, intended to replace the template pattern, that provides the base styles to render a card. It provides "slots" for calling components to inject content into, ie.:

```
<Card>
  <Card.Image src={url} />
  <Card.Info>
    Info here
  </Card.Info>
  <Card.Title>
    Title here
  </Card.Title>
<Card>
```

  It also provides a `Card.Content` slot that will occupy the full `Card` for the panel styles only.

- The `Card` component provides `link` prop for cards that are links in their entirety - this is in use by the Homepage "News and Events" cards.

- I added some adjustments to the carousel width calculations so the cards align with the page container width (1280px), though the cards are now fixed width (302px and 194px including borders), so do not expand to parent width and do not need widths applied by the carousel. These aren't working on mobile where they're squashed, though @ChristopherChudzicki is looking at the carousel at the moment so is in his good hands.

- The LearningResourceCard and LearningResourceCardTemplate are still in use for the list variants (by ListDetailsPage and SearchPage), so are left in the mit-open workspace. These will be refactored out as the list variant is implemented.

- This renames the MUI Card export from `ol-components` to `MuiCard`. I didn't want to call our card anything other than `Card` and am not sure we should relinquish all the good component names to MUI. We could prefix all the MUI exports and I see the benefit in having a single entrypoint for all components, though my preference would be not to re-export library components at all. I see greater benefit in it being clear in the consuming code that one of our components vs a library component is being imported and I would import directly from MUI - it's unlikely we'd add a new component library to the project unnoticed.
